crates/codegen/ebnf/src/ebnf_writer.rs

```diff
@@ -0,0 +1,76 @@
+#[derive(Debug)]
+#[allow(non_camel_case_types)]
+pub enum TokenKind {
+    comment,
+    constant,
+    keyword,
+    operator,
+    string,
+}
+
+pub trait EBNFWriter {
+    fn write_token(&mut self, kind: TokenKind, value: &str);
+    fn write_line_break(&mut self);
+
+    fn write_production_definition(&mut self, production_name: &str);
+    fn write_production_reference(&mut self, production_name: &str);
+    fn write_local_production_reference(&mut self, _parent_name: &str, production_name: &str) {
+        self.write_production_reference(production_name)
+    }
+
+    fn write_prelude(&mut self) {}
+    fn write_postlude(&mut self) {}
+    fn write_line_start(&mut self) {}
```

Comment:
In a next PR, Spec will change to no longer generate HTML directly, as it is really bad for diffing/reading. I'm working on a shorter syntax that can generate HTML at runtime in browser. For that, thoughts on:

1. replacing `write_production_definition()` with `start_definition()` and `finish_definition()`
2. removing `write_line_start()` and `write_line_break()`

Removing the concept of "lines" from writers, since not all of them will deal with whitespace/line breaks this way.